doc/designs/quic-design/quic-fault-injector.md

```diff
@@ -0,0 +1,507 @@
+test will be able to "listen" for a specifc event occuring and be informed about
+it when it does. Examples of events might include:
+
+- An EncryptedExtensions handshake message being sent
+- An ACK frame being sent
+- A Datagram being sent
+
+Each listener will be provided with additional data about the specific event.
+For example a listener that is listening for an EncryptedExtensions message will
+be provided with the parsed contents of that message in an easy to use
+structure. Additional helper functions will be provided to make changes to the
+message (such as to resize it).
+
+Initially listeners will only be able to listen for events on the server side.
+This is because, in MVP, it will be the client side that is under test - so the
+faults need to be injected into protocol elements sent from the server. Post
+MVP this will be extended in order to be able to test the server. It may be that
+we need to do this during MVP in order to be able to observe protocol elements
+sent from the client without modifying them (i.e. in order to confirm that the
+client is behaving as we expect). This will be added if required as we develop
+the tests.
+
+It is expected that the Fault Injector API will expand over time as new
+listeners and helper functions are added to support specific test scenarios. The
+initial API will provide a basic set of listeners and helper functions in order
+to provide the basis for future work.
+
+The following outlines an illustrative set of functions that will initially be
+provided. A number of `TODO(QUIC)` comments are inserted to explain how we
+might exand the API over time:
+
+```` C
+/* Type to represent the Fault Injector */
+typedef struct ossl_quic_fault OSSL_QUIC_FAULT;
+
+/*
+ * Structure representing a parse EncryptedExtension message. Listeners can
+ * make changes to the contents of structure objects as required and the fault
+ * injector will reconstruct the message to be sent on
+ */
+typedef struct ossl_qf_encrypted_extensions {
+    /* EncryptedExtension messages just have an extensions block */
+    unsigned char *extensions;
+    size_t extensionslen;
+} OSSL_QF_ENCRYPTED_EXTENSIONS;
+
+/*
+ * Given an SSL_CTX for the client and filenames for the server certificate and
+ * keyfile, create a server and client instances as well as a fault injector
+ * instance
+ */
+int qtest_create_quic_objects(SSL_CTX *clientctx, char *certfile, char *keyfile,
+                              QUIC_TSERVER **qtserv, SSL **cssl,
+                              OSSL_QUIC_FAULT **fault);
+
+/*
+ * Free up a Fault Injector instance
+ */
+void ossl_quic_fault_free(OSSL_QUIC_FAULT *fault);
+
+/*
+ * Run the TLS handshake to create a QUIC connection between the client and
+ * server.
+ */
+int qtest_create_quic_connection(QUIC_TSERVER *qtserv, SSL *clientssl);
+
+/*
+ * Confirm the server has received a protocol error
+ */
+int qtest_check_server_protocol_err(QUIC_TSERVER *qtserv);
```

Comment:
Would be nice to have a variant of this that confirms the correct QUIC Transport Error code is returned &mdash; we should be checking that.